File: src/main/java/com/EmosewaPixel/pixellib/resourceAddition/TagMaps.java
package com.EmosewaPixel.pixellib.resourceAddition;

import com.EmosewaPixel.pixellib.materialSystem.materials.IMaterialItem;
import com.google.common.collect.ListMultimap;
import com.google.common.collect.MultimapBuilder;
import net.minecraft.block.Block;
import net.minecraft.item.Item;

public class TagMaps {
    protected static ListMultimap<String, Item> ITEM_TAGS = MultimapBuilder.treeKeys().arrayListValues().build();
    protected static ListMultimap<String, Block> BLOCK_TAGS = MultimapBuilder.treeKeys().arrayListValues().build();

    public static void addMatItemToTag(IMaterialItem item) {
        if (item instanceof Item) {
            addItemToTag(item.getObjType().getName() + "s/" + item.getMaterial().getName(), (Item) item);
            addItemToTag(item.getObjType().getName() + "s", (Item) item);
            if (item.getMaterial().hasSecondName())
                addItemToTag(item.getObjType().getName() + "s/" + item.getMaterial().getSecondName(), (Item) item);
        }
        if (item instanceof Block) {
            addBlockToTag(item.getObjType().getName() + "s/" + item.getMaterial().getName(), (Block) item);
            addBlockToTag(item.getObjType().getName() + "s", (Block) item);
            if (item.getMaterial().hasSecondName())
                addBlockToTag(item.getObjType().getName() + "s/" + item.getMaterial().getSecondName(), (Block) item);
        }
    }

    public static void addItemToTag(String tag, Item item) {
        ITEM_TAGS.put(tag, item);
    }

    public static void addBlockToTag(String tag, Block block) {
        BLOCK_TAGS.put(tag, block);
        ITEM_TAGS.put(tag, Item.getItemFromBlock(block));
    }
}

File: src/main/kotlin/com/scientianova/versatile/resources/TagMaps.kt
@file:JvmName("TagMaps")

package com.scientianova.versatile.resources

import com.google.common.collect.ListMultimap
import com.google.common.collect.MultimapBuilder
import com.scientianova.versatile.common.extensions.toResLoc
import net.minecraft.block.Block
import net.minecraft.fluid.Fluid
import net.minecraft.item.Item
import net.minecraft.util.ResourceLocation

val ITEM_TAGS: ListMultimap<ResourceLocation, Item> = MultimapBuilder.treeKeys().arrayListValues().build()
val BLOCK_TAGS: ListMultimap<ResourceLocation, Block> = MultimapBuilder.treeKeys().arrayListValues().build()
val FLUID_TAGS: ListMultimap<ResourceLocation, Fluid> = MultimapBuilder.treeKeys().arrayListValues().build()

fun addItemToTag(tag: ResourceLocation, item: Item) = ITEM_TAGS.put(tag, item)
fun addItemToTag(tag: String, item: Item) = addItemToTag(tag.toResLoc(), item)

fun addBlockToTag(tag: ResourceLocation, block: Block) = BLOCK_TAGS.put(tag, block)
fun addBlockToTag(tag: String, block: Block) = addBlockToTag(tag.toResLoc(), block)

fun addFluidToTag(tag: ResourceLocation, fluid: Fluid) = FLUID_TAGS.put(tag, fluid)
fun addFluidToTag(tag: String, fluid: Fluid) = addFluidToTag(tag.toResLoc(), fluid)